faf-commons-data/src/main/java/com/faforever/commons/replay/ReplaySemantics.java

```diff
@@ -0,0 +1,152 @@
+   * @return
+   */
+  public static List<RegisteredEvent> registerEvents(List<Source> sources, List<Event> events) {
+    final AtomicInteger tick = new AtomicInteger(0);
+    final AtomicInteger commandSourceId = new AtomicInteger(-1);
+
+    return events.stream().map((event) -> switch (event) {
+      case Event.Advance(int ticksToAdvance) -> {
+        tick.addAndGet(ticksToAdvance);
+        yield null;
+      }
+
+      case Event.SetCommandSource (int playerIndex) -> {
+        commandSourceId.set(playerIndex);
+        yield null;
+      }
+
+      default -> new RegisteredEvent(tick.intValue(), sources.get(commandSourceId.intValue()), event);
+    }).filter(Objects::nonNull).toList();
+  }
+
+  /**
+   * Retrieves all events that are chat messages
+   *
+   * @param events A list of events
+   * @return A list of events that are chat messages
+   */
+  public static List<ChatMessage> findChatMessages(List<Source> sources, List<RegisteredEvent> events) {
+    return events.stream().map((registeredEvent) -> switch (registeredEvent.event()) {
+
+      // TODO: the fact that we piggy-back on the 'GiveResourcesToPlayer' callback to embed chat messages is all wrong! We should instead introduce an alternative callback with the sole purpose to send messages.
+      //  Requires refactoring in the game!
+
+      case Event.LuaSimCallback(
+        String func, LuaData.Table callbackTable, Event.CommandUnits commandUnits
+      ) when func.equals("GiveResourcesToPlayer") -> {
+        // TODO: this field has no meaning and can be manipulated, instead use the authorised command source.
+        //  Requires refactoring in the game!
+        if (!(callbackTable.value().get("From") instanceof LuaData.Number from)) {
+          yield null;
+        }
+
+        // focus army starts is 1-based instead of 0-based, to align it we subtract 1
+        if (from.value() - 1 <= -2) {
+          yield null;
+        }
+
+        // TODO: this field has no meaning and can be manipulated, instead use the authorised command source.
+        //  Requires refactoring in the game!
+        if (!(callbackTable.value().get("Sender") instanceof LuaData.String sender)) {
+          yield null;
+        }
+
+        // TODO: apparently all players create a sim callback that contains the chat message. This hack is how we skip it,
+        //   Requires refactoring in the game!
+        if (!Objects.equals(sender.value(), registeredEvent.source().name())) {
+          yield null;
+        }
+
+        if (!(callbackTable.value().get("Msg") instanceof LuaData.Table msgTable)) {
+          yield null;
+        }
+
+        // TODO: this is 1 out of the 2 legitimate fields
+        if (!(msgTable.value().get("to") instanceof LuaData.String msgTo)) {
+          yield null;
+        }
+
+        // TODO: this is 2 out of the 2 legitimate fields
+        if (!(msgTable.value().get("text") instanceof LuaData.String msgText)) {
+          yield null;
+        }
```

Comment:
Technically for these you could use the full record pattern. So LuaData.Number(float value)

Just to remove all the calls of .value

Reply:
Processed in https://github.com/FAForever/faf-java-commons/pull/143/commits/1b9492f10246b50436790dec2539c836a80ffabd